gax/src/main/java/com/google/api/gax/batching/v2/BatcherImpl.java

```diff
@@ -0,0 +1,268 @@
+      Builder<EntryT, ResultT, RequestT, ResponseT> newBuilder() {
+    return new Builder<>();
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public ApiFuture<ResultT> add(final EntryT entry) {
+    Preconditions.checkState(!isShutDown, "Cannot perform batching on a closed connection");
+    lock.lock();
+    try {
+      flowController.reserve(entry);
+
+      boolean anyThresholdReached = isAnyThresholdReached(entry);
+      if (entries == null) {
+        entries = descriptor.getRequestBuilder();
+        resultFutures = new ArrayList<>();
+
+        // Scheduling a job with maxDelay, after each entries assignment.
+        if (!anyThresholdReached) {
+          executor.schedule(flushCurrentBatchRunnable, maxDelay.toMillis(), TimeUnit.MILLISECONDS);
+        }
+      }
+      entries.add(entry);
+      SettableApiFuture<ResultT> result = SettableApiFuture.create();
+      resultFutures.add(result);
+      ApiFutures.addCallback(
+          result,
+          new ApiFutureCallback<ResultT>() {
+            @Override
+            public void onFailure(Throwable t) {
+              flowController.release(entry);
+            }
+
+            @Override
+            public void onSuccess(ResultT result) {
+              flowController.release(entry);
+            }
+          },
+          directExecutor());
+
+      if (anyThresholdReached) {
+        flush();
+      }
+
+      return result;
+    } catch (FlowController.FlowControlException e) {
+      throw BatchingException.fromFlowControlException(e);
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public void flush() {
+    lock.lock();
+    try {
+      if (entries != null) {
+        final RequestT request = entries.build();
+        entries = null;
+
+        final List<SettableApiFuture<ResultT>> results = resultFutures;
+        resultFutures = null;
+
+        final ApiFuture<ResponseT> responseFuture = callable.futureCall(request);
+        responses.add(responseFuture);
+        ApiFutures.addCallback(
+            responseFuture,
+            new ApiFutureCallback<ResponseT>() {
+              @Override
+              public void onSuccess(ResponseT response) {
+                responses.remove(responseFuture);
+                descriptor.splitResponse(response, results);
+              }
+
+              @Override
+              public void onFailure(Throwable throwable) {
+                responses.remove(responseFuture);
+                descriptor.splitException(throwable, results);
+              }
+            },
+            directExecutor());
+        resetThresholds();
+      }
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public void close() {
+    shutdown();
+  }
+
+  @Override
+  public void shutdown() {
+    isShutDown = true;
+    flush();
+    if (!responses.isEmpty()) {
+      executor.shutdown();
+      try {
+        executor.awaitTermination(60, TimeUnit.SECONDS);
```

Comment:
@ajaaym I did add the `awaitTermination` inside `#shutdown` operation. 
Is it ok to block the execution even for shutdown?... I am not sure as existing batch never had any `awaitTermination`

Reply:
not for shutdown, only for close.